### PR TITLE
Add more Exception messages for Config.getValue()

### PR DESCRIPTION
**Pull request by Joseph-Cass**
Partly addresses #485
In particular, resolves:

- https://github.com/smallrye/smallrye-config/issues/485#issuecomment-762208755
- https://github.com/smallrye/smallrye-config/issues/485#issuecomment-772418418

Under the current implementation all of the tests from `ConfigValueConversionRulesExceptionsTest.java` would have thrown
```
Property %s is required but the value was not found or is empty
```
Except from `badConversion()` which would have thrown:
```
Expected an integer value, got "notInt"
```

### CODE REVIEW

**Comment by radcortez** [comment]
@Joseph-Cass Is this ready?

**Comment by Joseph-Cass** [comment]
[This comment](https://github.com/smallrye/smallrye-config/blob/59a74bd10ffd2d6908375b0b117f67900c5bf47d/implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java#L142-L149) may change depending on the outcome of [this discussion](https://github.com/eclipse/microprofile-config/issues/675), but other than that, I'm happy if you are :)

File: implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java [hunk 1]
```diff
@@ -139,20 +139,44 @@
             }
         }
 
-        String value = configValue.getValue();
+        /*
```
Comment (radcortez) [inline implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java:142]: May be more interesting to move this to a java doc section.

File: implementation/src/main/java/io/smallrye/config/ConfigMessages.java [hunk 2]
```diff
@@ -57,7 +57,8 @@
     @Message(id = 13, value = "No Converter registered for %s")
     IllegalArgumentException noRegisteredConverter(Class<?> type);
 
-    @Message(id = 14, value = "Property %s is required but the value was not found or is empty")
+    // Returns a String rather than a NoSuchElementException for a slight performance improvement as throwing this exception could be quite common.
+    @Message(id = 14, value = "The config property %s is required but it could not be found in any config sources")
```
Comment (radcortez) [inline implementation/src/main/java/io/smallrye/config/ConfigMessages.java:61]: Maybe `in any config source` (drop plural)?

**Review by radcortez — CHANGES_REQUESTED** [review]

File: implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java (hunk 1 above)
Comment (Joseph-Cass) [reply implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java]: Is this what you meant? 
https://github.com/smallrye/smallrye-config/blob/39c034bf374d2de98ca066294a72722c185a0821/implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java#L128-L142

**Comment by Joseph-Cass** [comment]
Does this look okay to you @radcortez ?

File: implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java (hunk 1 above)
Comment (radcortez) [reply implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java]: Yes.

**Comment by radcortez** [comment]
Yes! Thank you!

**Review by radcortez — APPROVED** [review]

**Comment by radcortez** [comment]
Cannot backport to 1.x due to merge conflicts. Please backport manually:
Run:
```
git clone https://github.com/smallrye/smallrye-config.git
git fetch origin pull/504/head:pr-504
git checkout -b 1.x origin/1.x
git checkout -b backport-#504-to-1.x
# One or more of the following command will fail, you will need to fix the conflict manually
git cherry-pick ed53f84b2ba931cb9b217f1dc61373fb37465c45
git cherry-pick d091ce36e7784cec833058328684330848e4e5a2
git cherry-pick ac2c83fe02dea7543bca7e2b42e0830bd09f2370
# Once all commits have been cherry-picked:
git push --set-upstream origin backport-#504-to-1.x
```
To fix the conflict, first check which file is impacted using: `git status`
For each file with a resolved conflict, execute: `git add $FILE`
Then, commit the files using the same commit message as the original commit: `git commit -m "..."`

Once done and pushed, open the pull request.

* Title: [1.x] Backport Add more Exception messages for Config.getValue()
* Message: Backport #504 to 1.x.
* ⚡ **Set the target branch to 1.x** 
* Set the milestone and the labels if needed

**Comment by Joseph-Cass** [comment]
@radcortez I'm not sure this PR is suitable for 1.x. It updates error messages related to "empty" and null behaviour which changed in the spec between MP Config 1.x -> 2.0?

**Comment by radcortez** [comment]
Cannot backport to 1.x due to merge conflicts. Please backport manually:
Run:
```
git clone https://github.com/smallrye/smallrye-config.git
git fetch origin pull/504/head:pr-504
git checkout -b 1.x origin/1.x
git checkout -b backport-504-to-1.x
# One or more of the following command will fail, you will need to fix the conflict manually
git cherry-pick ed53f84b2ba931cb9b217f1dc61373fb37465c45
git cherry-pick d091ce36e7784cec833058328684330848e4e5a2
git cherry-pick ac2c83fe02dea7543bca7e2b42e0830bd09f2370
# Once all commits have been cherry-picked:
git push --set-upstream origin backport-504-to-1.x
```
To fix the conflict, first check which file is impacted using: `git status`
For each file with a resolved conflict, execute: `git add $FILE`
Then, commit the files using the same commit message as the original commit: `git commit -m "..."`

Once done and pushed, open the pull request.

* Title: [1.x] Backport Add more Exception messages for Config.getValue()
* Message: Backport #504 to 1.x.
* ⚡ **Set the target branch to 1.x** 
* Set the milestone and the labels if needed

**Comment by radcortez** [comment]
Ahah, I was actually using this as a test for a script I wrote to do backports. 

There are some interesting things that I would like to backport, but I haven't really tried it yet.

**Pull request merged**